docs/adr/0006-rapid-releases-and-moving-towards-a-monorepo.md

```diff
@@ -0,0 +1,123 @@
+  * Bad, because we need implement new automation.
+  * Bad, because when breaking changes cause the automatic nightly merge to fail for an application.  This means a developer will get the relatively urgent task of resolving things.
+  * Good, because this avoids the current state of affairs where our `main` branch is not ready to be taken by the applications.  This means that a) our current code is not getting used and tested and b) it's very complicated to get other changes to that application, including bugfixes.
+  * Bad, because breaking changes require the corresponding PRs to be approved from all other teams before they can be merged into application-services.
+  * Good, because this is how things will work in a monorepo world.  Adjusting our workflow now prepares us for the future and will give us insights that may help guide the monorepo migration.
+  * Good, because our version numbers will align more closely with the rest of Mozilla.
+  * Good, because it makes it easier to know which application-services version the applications are using.  The current Firefox X release will always be using the latest application-services.X release -- for both the mobile apps and Desktop.
+
+# Non-normative discussion
+
+## Impact on other teams
+### Application teams -- Android, iOS, and Desktop
+
+* Phase 1
+  * The application-services version number style will change
+  * There may be a increase in the amount of application-service version upgrade PRs to approve.  We'll want to make sure that each application is on an application-services release that matches their release (i.e. avoiding having Fenix 110.x on application-services 109.x).  For the Android/iOS teams this will only be a small increase, but for Desktop it will be larger.
+* Phase 2
+  * More application-services PRs to approve, especially at the start.  Eventually, we should be able to arrange it so most PRs get auto-merged.
+
+### Release Management
+
+* Phase 1
+  * Will now handle tagging, releasing, and version bumping application-services at the end of the nightly cycle.
+  * Will now handle patch releases from the release / nightly branches
+* Phase 2
+  * Maybe handle backing out changes when the auto-merge fails?  I'm not sure who should own this.
+
+### Release Engineering
+
+* Support the application-services team in developing the automation.  I believe most of the proposed changes are simple enough that application-services team can handle it ourselves.
+
+## Nightly releases
+
+The ADR mentions making nightly releases from the main branch.  There are two basic approaches for this:
+
+  * **Create new releases**: Create a new version number, GitHub tag, GitHub release, Kotlin/swift packages, etc.
+  * **Update a single release**.  Use a name like `[firefox-version].0a1` for the nightlies and don't change it.  Don't create GitHub tags or releases.  Re-publish the existing Kotlin/Swift packages rather than create new ones.
```

Comment:
IMO, Nightlies shouldn't be getting tagged as releases. I would think we'd be doing something akin to how we ship new GeckoView nightlies every day where they get uploaded to the Nightly Maven repo with a [version].[date] numbering scheme, i.e. 110.0.20230105143224 as shown in https://nightly.maven.mozilla.org/?prefix=maven2/org/mozilla/components/browser-engine-gecko/. Then the first tagged release for a given train would be after the .0 release is branched and shipped.

Reply:
This makes a lot of sense to me as well.  The only reason for tagging them is that the existing automation creates the PRs when there's a new release, so maybe it simplifies the implementation.  Hopefully we can modify the automation to do what you suggest though.

Reply:
I'll add that no tags or releases for nightlies is consistent with what we do elsewhere, too. (Eg: Fenix uses [Taskcluster's cron scheduling](https://github.com/mozilla-mobile/fenix/blob/main/.cron.yml) for nightlies.)